### DISCUSSION TITLE
Errors when I try to quickly restart Speechly

### DISCUSSION THREAD

**Discussion by Lex-Kent** [speechly/speechly#249] [General]
Hi all,

I'm creating a chatbot using Speechly and I've encountered an issue when I try and start a new Speechly instance before the previous one has closed.

My app works really well when my bot talks and then asks a question, and I start Speechly to listen for a response, then stop the Speechly instance when my user has answered using `segment.isFinal`.

However, I'd now like to allow the user to interrupt the bot by saying something like "Stop, repeat that last thing." or "Please skip this part." or whatever. So I'd like to stop the previous Speechly listener when the user has replied, and then launch a new one which will listen for interruptions (it will no doubt hear my bot speaking, too, but I haven't got that far yet!). I have a function called `stopSpeechlyListening`, which when called with a restart argument, should `restart` Speechly as soon as it can:

```
ctx.stopSpeechlyListening = async (restart) => {
  ctx.isCharacterListening = false;
  if (ctx.speechly.isActive()){
    await ctx.speechly.stop();
    ctx.listenFinished();
    if (restart){
      ctx.startListening();
    }
  }
  else{
    if (restart){
      ctx.startListening();
    }
  }
}
```

However, when I call the function with `ctx.stopSpeechlyListening(true)`, I often see the following error:

`Uncaught (in promise) Error: [Decoder] Unable to complete startContext: Expected Connected state, but was in Active.`

and my app fails.

I also noticed that when this happens, even though `ctx.speechly.isActive()` returns `false`, `ctx.speechly.decoder.state` returns `3` (`Active`).

Does anyone have any advice please?

Many thanks,

**Comment by bigdatabaracus** [comment]
Hello @Lex-Kent, thanks for reaching out. @arzga From the Speechly team would probably be the best person to help you out here but he is current out-of-office. He'll be back next week. 👍

best regards, -Hannes-

**Comment by arzga** [comment]
Hi @Lex-Kent,

First of all, please ensure that you're using the latest version (2.6.1 as of now) of @speechly/browser-client.

Speechly API guarantees that if you `await BrowserClient.start()` or `await BrowserClient.stop()` you will get back the audio context id string which you can use to match the transcription to certain piece of speech.

When the engine is already listening (`BrowserClient.isActive()` returns `true`), and a `start()` attempt is made, an exception is thrown. This is due to the reason we can't provide a new audio context id as Speechly is already listening and processing an earlier audio snippet.

Glancing thru your code I notice that you have an `isActive()` check in place for stopping listening. However, I don't see the exact contents of `ctx.startListening()` or other places that might call start and stop. To ensure that throwing never happens, may I suggest that you wrap all `BrowserClient.stop()` and `BrowserClient.start()` calls with appropriate checks like so:

Safe stop:
```
if (ctx.speechly.isActive()) {
  await ctx.speechly.stop();
}
```

Safe start:
```
if (!ctx.speechly.isActive()) {
  await ctx.speechly.start();
}
```

Of course you might want to check why start/stop calls go out-of-sync to ensure no other funny stuff happens 😁 . Please also ensure that you are not using `start()` and `stop()` calls when Speechly's automatic voice activity detection (VAD) is enabled, as they may interfere with each other.

FYI, `start()` and `stop()` flip the `isActive()` state immediately whereas the internal state reported by `onStateChange` callback (or `ctx.speechly.decoder.state`) only changes when all related packets from the backend have been received. It is safe to call `stop()` and `start()` in alternating order immediately after each other (even without awaiting) as there is an internal queue for the requests.

**Reply by Lex-Kent** [reply]
@bigdatabaracus @arzga 

Thank you both for taking the time to reply.

@arzga I still don't really know what was wrong with my code, but I've decided to tackle this differently. I now invoke Speechly only once, with `vad` enabled. I have created a flag called `botIsTalking` that's set to true when my bot is speaking and another `awaitingUserReply` which is set to true when the bot asks a question. This looks to be working really well - `vad` is handling the stopping and starting of Speechly automatically (presumably using magic).

So I'm now able to tell when the user is attempting to interrupt my bot and it is working remarkably well. As a matter of fact, it is working _too_ well - Speechly never seems to hear my bot's dialogue even when I have my PC speakers turned up all the way. My next challenge is to understand why that is and to deal with it if and when I can make it happen!
